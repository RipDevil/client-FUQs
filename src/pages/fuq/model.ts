export interface FuqType {
  _id: string;
  likes: number;
  _lastEditor: string;
  pending: boolean;
  title: string;
  text: string;
  crdate: string;
};
