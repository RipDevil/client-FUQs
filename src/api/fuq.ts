import { AxiosError } from 'axios';
import { useQuery } from 'react-query';

import { $config } from 'config/model';
import { call } from './lib';
import { FuqType } from 'pages/fuq/model';

/**
 * Can return a single FUQ randomly or not if the ID param is presented
 * @param id - id of a FUQ
 */
export const useSingleFuq = (id?: string) => {
  const { server } = $config.getState();

  return useQuery<unknown, AxiosError, FuqType>(`Get ${id ? 'random' : 'single'} FUQ`, () =>
    call(`${server}/fuq${id ? '/' + id : ''}`, 'GET'),
  );
};
