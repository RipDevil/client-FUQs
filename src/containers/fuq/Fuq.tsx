import * as React from 'react';
import { RouteComponentProps } from 'react-router-dom';

import Layout from 'containers/layout/Layout';

export interface FuqMatchParams {
  id: string;
}

const Fuq: React.SFC<RouteComponentProps<FuqMatchParams>> = (props) => {
  console.log('props :>> ', props);
  const {
    match: {
      params: { id },
    },
  } = props;

  return (
    <Layout>
      <b>/fuq/{id}</b>
    </Layout>
  );
};

export default Fuq;
