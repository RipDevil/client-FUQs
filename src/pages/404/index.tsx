import * as React from 'react';

import { Layout } from 'templates';

const PageNotFound: React.FC = () => {
  return (
    <Layout>
      <b>/404</b>
    </Layout>
  );
};

export default PageNotFound;
