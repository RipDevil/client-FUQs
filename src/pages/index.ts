import SingleFuq from './fuq';
import Admin from './admin';
import Login from './login';

import PageNotFound from './404';

export const routes = () => [
  {
    path: '/',
    exact: true,
    component: SingleFuq,
  },
  {
    path: '/fuq/:id',
    exact: true,
    component: SingleFuq,
  },
  {
    path: '/badmin',
    exact: true,
    component: Admin,
  },
  {
    path: '/login',
    exact: true,
    component: Login,
  },
  { component: PageNotFound },
];
