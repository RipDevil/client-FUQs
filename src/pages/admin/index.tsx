import * as React from 'react';
import { Redirect } from 'react-router-dom';
import { useStore } from 'effector-react';
import { MenuInfo } from 'rc-menu/lib/interface';
import { Menu, Col, Button, message } from 'antd';

import { AdminLayout } from 'templates';
import { UsersGrip, FuqsGrip } from 'components/admin';
import { $auth, authReset } from 'pages/login/model';
import { useLogout } from 'api';

const { Item } = Menu;

type ItemType = {
  key: number;
  title: string;
};

const DEFAULT_GRIP: string = '0';
const Admin: React.FC = () => {
  const [grip, setGrip] = React.useState<string>('Users');
  const authStore = useStore($auth);
  const [items] = React.useState<ItemType[]>([
    {
      key: 0,
      title: 'Users',
    },
    {
      key: 1,
      title: 'FUQs',
    },
  ]);

  const { mutate: logout, isError, isLoading, isSuccess } = useLogout();

  React.useEffect(() => {
    isError && message.warning('Can not log off');
  }, [isError]);

  React.useEffect(() => {
    isSuccess && authReset();
  }, [isSuccess]);

  const onItemClick = ({ key: reactKey }: MenuInfo) => {
    const key = reactKey.toString();
    setGrip(items[parseInt(key)].title);
  };

  // @mna: is this the best variant?
  if (authStore.refreshToken === '' || authStore.refreshToken === '') {
    return <Redirect to="/login" />;
  }

  return (
    <AdminLayout>
      <Col span={5}>
        <Button data-testid="logout-btn" onClick={logout} type="primary" loading={isLoading} block>
          Logout
        </Button>
        <Menu onClick={onItemClick} mode="inline" defaultSelectedKeys={[DEFAULT_GRIP]} style={{ height: '100%' }}>
          {items && items.length ? (
            items.map((item) => (
              <Item data-testid={item.key} key={item.key}>
                {item.title}
              </Item>
            ))
          ) : (
            <b>...</b>
          )}
        </Menu>
      </Col>
      <Col span={19} style={{ padding: '5px' }}>
        {grip === items[0].title && <UsersGrip />}
        {grip === items[1].title && <FuqsGrip />}
      </Col>
    </AdminLayout>
  );
};

export default Admin;
