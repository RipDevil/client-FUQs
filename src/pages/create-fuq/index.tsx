import * as React from 'react';
import { RouteComponentProps, useHistory } from 'react-router-dom';
import { message } from 'antd';
import { History as H } from 'history';

import { Layout } from 'templates';
import { usePutFuq } from 'api';
import { Spinner } from 'components/common';
import { CreateFuqCard, NewFuqFormTypes } from 'components/common/createFuqCard/create-fuq-card';

export interface CreateFuqMatchParams {
  id?: string;
}

const CreateFuq: React.FC<RouteComponentProps<CreateFuqMatchParams>> = (props) => {
  const { id } = props.match.params;

  const { mutateAsync: createFuq, isError, isLoading, isSuccess, data } = usePutFuq();
  const history = useHistory<H>();

  React.useEffect(() => {
    if (isSuccess && data?._id) {
      message.success({
        content: 'Success!',
      });

      history.push(`/fuq/${data._id}`);
    }
  }, [data, isSuccess]);

  React.useEffect(() => {
    isError &&
      message.error({
        content: 'Error!',
      });
  }, [isLoading, isError]);

  function onFormSubmit(values: NewFuqFormTypes) {
    createFuq(values);
  }

  return (
    <Layout>
      {isLoading && <Spinner transparent={false} text="Adding a new FUQ, mi' lad" />}
      <CreateFuqCard onSubmit={onFormSubmit} influencer={id} />
    </Layout>
  );
};

export default CreateFuq;
