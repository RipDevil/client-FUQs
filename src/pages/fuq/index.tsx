import * as React from 'react';
import { RouteComponentProps } from 'react-router-dom';

import { Layout } from 'templates';
import { Spinner, FuqCard } from 'components/common';
import { useSingleFuq } from 'api';

export interface FuqMatchParams {
  id?: string;
}

export const SingleFuq: React.FC<RouteComponentProps<FuqMatchParams>> = (props) => {
  const { isFetching, data: fuq, isSuccess, isError, error } = useSingleFuq(props.match.params.id);

  return (
    <Layout>
      {isError && error && <FuqCard text={error.message} title="Error" id={props.location.pathname} />}
      {isFetching && <Spinner text="Loading a FUQ" />}
      {isSuccess && fuq?._id && <FuqCard text={fuq?.text || ''} title={fuq?.title || ''} id={fuq._id} />}
    </Layout>
  );
};
