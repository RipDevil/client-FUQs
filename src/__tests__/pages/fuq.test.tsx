import { QueryClient, QueryClientProvider } from 'react-query';
import { Router, Route } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { render, waitFor, cleanup } from '@testing-library/react';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import { SingleFuq as FuqPage } from 'pages/fuq';

let queryClient: any;
let mockAxios: any;

beforeAll(() => {
  queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        refetchOnWindowFocus: false,
        retry: false,
      },
    },
  });
  mockAxios = new MockAdapter(axios);
});

afterEach(() => {
  mockAxios.reset();
  cleanup();
});

describe('Config wrapper tests', () => {
  it('Should render loader', async () => {
    mockAxios.onGet('/fuq').reply(200, {});

    const { getByText } = render(
      <QueryClientProvider client={queryClient}>
        <Router history={createMemoryHistory({ initialEntries: ['/'] })}>
          <Route path={'/'} component={FuqPage} />
        </Router>
      </QueryClientProvider>,
    );

    const loadingElement = getByText('Loading a FUQ');
    expect(loadingElement).toBeInTheDocument();
    expect(document.body).toMatchSnapshot();
  });

  it('Should render FUQ', async () => {
    mockAxios.onGet('/fuq').reply(200, {
      _id: '13',
      likes: 0,
      _lastEditor: 'string',
      pending: false,
      title: 'string',
      text: 'string',
      crdate: 'string',
    });

    const { getAllByText } = render(
      <QueryClientProvider client={queryClient}>
        <Router history={createMemoryHistory({ initialEntries: ['/'] })}>
          <Route path={'/'} component={FuqPage} />
        </Router>
      </QueryClientProvider>,
    );

    const resolvedElements = await waitFor(() => getAllByText('string'));
    resolvedElements.forEach((item) => {
      expect(item).toBeInTheDocument();
    });
    expect(document.body).toMatchSnapshot();
  });

  it('Should render FUQ with params', async () => {
    mockAxios.onGet('/fuq/13').reply(200, {
      _id: '13',
      likes: 0,
      _lastEditor: 'string',
      pending: false,
      title: 'string',
      text: 'string',
      crdate: 'string',
    });

    const { getAllByText } = render(
      <QueryClientProvider client={queryClient}>
        <Router history={createMemoryHistory({ initialEntries: ['/fuq/13'] })}>
          <Route path={'/fuq/:id'} component={FuqPage} />
        </Router>
      </QueryClientProvider>,
    );

    const resolvedElements = await waitFor(() => getAllByText('string'));
    resolvedElements.forEach((item) => {
      expect(item).toBeInTheDocument();
    });
    expect(document.body).toMatchSnapshot();
  });

  it('Should render error', async () => {
    mockAxios.onGet('/fuq/13').reply(404);

    const { getByText } = render(
      <QueryClientProvider client={queryClient}>
        <Router history={createMemoryHistory({ initialEntries: ['/fuq/13'] })}>
          <Route path={'/fuq/:id'} component={FuqPage} />
        </Router>
      </QueryClientProvider>,
    );

    const resolvedElement = await waitFor(() => getByText('Error'));
    expect(resolvedElement).toBeInTheDocument();
    expect(document.body).toMatchSnapshot();
  });
});
