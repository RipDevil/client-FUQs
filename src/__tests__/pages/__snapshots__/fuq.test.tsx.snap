// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config wrapper tests Should render FUQ 1`] = `
<body>
  .c0 {
  padding-top: 40vh;
  text-align: center;
}

.c1 {
  font-size: 50px !important;
}

<div>
    <div
      class="ant-row ant-row-center ant-row-middle c0"
    >
      <div
        class="ant-col"
        title="string"
      >
        <div
          class="ant-row ant-row-center"
        >
          <h1
            class="ant-typography ant-typography-danger ant-typography-ellipsis ant-typography-ellipsis-single-line c1"
          >
            string
          </h1>
        </div>
        <div
          class="ant-row ant-row-center"
        >
          <h3
            class="ant-typography ant-typography-ellipsis ant-typography-ellipsis-multiple-line"
          >
            string
          </h3>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Config wrapper tests Should render FUQ with params 1`] = `
<body>
  .c0 {
  padding-top: 40vh;
  text-align: center;
}

.c1 {
  font-size: 50px !important;
}

<div>
    <div
      class="ant-row ant-row-center ant-row-middle c0"
    >
      <div
        class="ant-col"
        title="string"
      >
        <div
          class="ant-row ant-row-center"
        >
          <h1
            class="ant-typography ant-typography-danger ant-typography-ellipsis ant-typography-ellipsis-single-line c1"
          >
            string
          </h1>
        </div>
        <div
          class="ant-row ant-row-center"
        >
          <h3
            class="ant-typography ant-typography-ellipsis ant-typography-ellipsis-multiple-line"
          >
            string
          </h3>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Config wrapper tests Should render error 1`] = `
<body>
  .c0 {
  padding-top: 40vh;
  text-align: center;
}

.c1 {
  font-size: 50px !important;
}

<div>
    <div
      class="ant-row ant-row-center ant-row-middle c0"
    >
      <div
        class="ant-col"
        title="Request failed with status code 404"
      >
        <div
          class="ant-row ant-row-center"
        >
          <h1
            class="ant-typography ant-typography-danger ant-typography-ellipsis ant-typography-ellipsis-single-line c1"
          >
            Error
          </h1>
        </div>
        <div
          class="ant-row ant-row-center"
        >
          <h3
            class="ant-typography ant-typography-ellipsis ant-typography-ellipsis-multiple-line"
          >
            Request failed with status code 404
          </h3>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Config wrapper tests Should render loader 1`] = `
.c0 {
  padding-top: 40vh;
  text-align: center;
}

.c2 {
  position: fixed;
  top: 50%;
  left: 50%;
  z-index: 1000;
  margin: -50px 0px 0px -50px;
  max-width: 140px;
}

.c1 {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: 999;
  background-color: rgba(253,253,253,1);
}

.c3 {
  text-align: center;
  margin-top: 1rem;
}

<body>
  <div>
    <div
      class="ant-row ant-row-center ant-row-middle c0"
    >
      <div
        class="c1"
      />
      <div
        class="c2"
      >
        <div>
          <i
            class="fas fa-10x fa-spin fa-cog"
          />
        </div>
        <div
          class="c3"
        >
          <h3
            class="ant-typography ant-typography-danger"
          >
            Loading a FUQ
          </h3>
        </div>
      </div>
    </div>
  </div>
</body>
`;
